satpy/multiscene.py

```diff
@@ -490,8 +521,56 @@ def _simple_frame_compute(self, writers, frame_keys, frames_to_write):
                 w = writers[frame_key]
                 w.append_data(product_frame.compute())
 
+    def _get_writers_and_frames(
+            self, filename, datasets, fill_value, ignore_missing,
+            enh_args, imio_args):
+        """Get writers and frames.
+
+        Helper function for save_animation.
+        """
+        scene_gen = self._scene_gen
+        writers = {}
+        frames = {}
+
+        scene_gen = self._scene_gen
+        first_scene = self.first_scene
+        scenes = iter(self._scene_gen)
+        info_scenes = [first_scene]
+        if 'end_time' in filename:
+            # if we need the last scene to generate the filename
+            # then compute all the scenes so we can figure it out
+            log.debug("Generating scenes to compute end_time for filename")
+            scenes = list(scenes)
+            info_scenes.append(scenes[-1])
+
+        available_ds = [first_scene.datasets.get(ds) for ds in first_scene.wishlist]
+        available_ds = [DatasetID.from_dict(ds.attrs) for ds in available_ds if ds is not None]
+        dataset_ids = datasets or available_ds
+
+        if not dataset_ids:
+            raise RuntimeError("No datasets found for saving (resampling may be needed to generate composites)")
+
```

Comment:
This function is quite long. Could it be split here? the first part could be called eg `filter_datasets`.

Reply:
It's already the result of splitting `save_animation`.  I started out by trying to split it at the start of the for loop, but then realised I needed to pass so much information to the split part that wasn't used in the caller that it became simpler to include the introductory part in the split-off function.  The tradeoff here is that if I split it further, the split part will have a large number of input and output arguments.  I haven't figured out a way to split this functionality in functions with a small number of arguments.

Reply:
If I split it here, then the next part needs `dataset_ids`, `scene_gen` (which is just `self._scene_gen`), `info_scenes`, `filename`, `fill_value`, `ignore_missing`, `enh_args`, `imio_args`, and I'd not know if passing 8 arguments to another small function is any clearer than the current function (which is indeed quite long).  I'm open to concrete suggestions that avoid this, in particular from @djhoese who authored this code (I just refactored it a bit).

Reply:
Ok, I think you are right, looks like a bigger part of the module would need to be refactored. So if we want to keep this small, we'll just have to live with the big function and the large number of arguments.